content/en/docs/Reference/file-based-catalogs.md

```diff
@@ -0,0 +1,586 @@
+
+## Specification
+
+### Structure
+
+File-based catalogs can be stored and loaded from directory-based filesystems.
+
+`opm` loads the catalog by walking the root directory and recursing into subdirectories. It attempts to load every file
+it finds and fails if any errors occur.
+
+Non-catalog files can be ignored using `.indexignore` files, which behave identically to `.gitignore` files. That is,
+they have the same rules for [patterns](https://git-scm.com/docs/gitignore#_pattern_format) and precedence.
+
+> #### Example `.gitignore` file
+> ```gitignore
+> # Ignore everything except non-object .json and .yaml files
+> **/*
+> !*.json
+> !*.yaml
+> **/objects/*.json
+> **/objects/*.yaml
+> ```
+
+
+Index maintainers have flexibility to chose their desired layout, but the OLM team recommends storing each package's
+file-based catalog blobs in separate sub-directories. Each individual file can be either JSON or YAML -- it is not
+necessary for every file in an index to use the same format. 
+
+This layout has the property that each sub-directoriy in the directory hierarchy is a self-contained index, which makes
+index composition, discovery, and navigation as simple as trivial filesystem operations.
+
+> #### Basic recommended structure
+> ```
+> index
+> ├── pkgA
+> │   └── index.yaml
+> ├── pkgB
+> │   ├── .indexignore
+> │   ├── index.yaml
+> │   └── objects
+> │       └── pkgB.v0.1.0.clusterserviceversion.yaml
+> └── pkgC
+>     └── index.json
+> ```
+
+This `index` could also be trivially included in a parent index by simply copying it into the parent index's root
+directory.
+
+### Schema
+
+At its core, file-based catalogs use a simple format that can be extended with arbitrary schemas. The format that all
+file-based catalog blobs must adhere to is the `Meta` schema. The below [cue][cuelang-spec] `_Meta` schema defines all
+file-based catalog blobs.
+
+> **NOTE**: No cue schemas listed in this specification should be considered exhaustive. The `opm validate` command has
+> additional validations that are difficult/impossible to express concisely in cue.
+
+```cue
+_Meta: {
+  // schema is required and must be a non-empty string
+  schema: string & !=""
+
+  // package is optional, but if it's defined, it must be a non-empty string
+  package?: string & !=""
+
+  // properties is optional, but if it's defined, it must be a list of 0 or more properties
+  properties?: [... #Property]
+}
+
+#Property: {
+  // type is required
+  type: string & !=""
+
+  // value is required, and it must not be null
+  value: !=null
+}
+```
+
+### OLM-defined schemas
+
+An OLM index currently uses two schemas: `olm.package` and `olm.bundle`, which correspond to OLM's existing package and
+bundle concepts.
+
+Each operator package in an index requires exactly one `olm.package` blob and one or more `olm.bundle` blobs.
+
+> **NOTE**: All `olm.*` schemas are reserved for OLM-defined schemas. Custom schemas must use a unique prefix (e.g. a
```

Comment:
Is there validation logic that checks for this in opm?

Reply:
For the `olm.*` properties, we expect their values to have a certain schema and we'll fail while parsing or during validation if those properties don't match what we expect.

We don't however have any inverse checks for "we see an olm.* property that we don't recognize". I'd be hesitant to add such a check though because then there would be version skew concerns between the `opm` binary and the set of known `olm.*` properties.